packages/runtime/src/state/react-page.ts

```diff
@@ -367,6 +374,52 @@ export function getElementTreeTranslatableData(
   return translatableData
 }
 
+export type TranslatableData = Record<string, Documents.Data>
+
+export type MergeTranslatableDataContext = {
+  translatableData: TranslatableData
+  mergeTranslatableData: (node: Documents.Element) => Documents.Element
+}
+
+export function mergeElementTreeTranslatableData(
+  state: State,
+  elementTree: Documents.ElementData,
+  translatableData: TranslatableData,
+): Documents.Element {
+  function mergeElementTreeNode(state: State, translatableData: TranslatableData) {
+    return function (node: Documents.Element): Documents.Element {
+      if (Documents.isElementReference(node)) return node
+
+      const elementDescriptors = getPropControllerDescriptors(state)
+      const descriptors = elementDescriptors.get(node.type)
+
+      if (descriptors == null) {
+        throw new Error(`Can't merge element of type "${node.type}" because it has no descriptors`)
+      }
+
+      const context: MergeTranslatableDataContext = {
+        translatableData,
+        mergeTranslatableData: mergeElementTreeNode(state, translatableData),
+      }
+      const mergedProps = {} as Record<string, Documents.Data>
+
+      for (const propName of Object.keys(descriptors)) {
+        const descriptor = descriptors[propName]
+
+        mergedProps[propName] = mergeTranslatableData(
+          descriptor,
+          `${node.key}:${propName}`,
+          node.props[propName],
+          context,
+        )
```

Comment:
In here is where I think we should be doing `mergeTranslatableData(descriptor, prop, translatedData, context)`

Reply:
`mergeTranslatedData` data is normalized with keys that include data referenced by element and propname
`copy` data is _normalized_ into a replacement context that includes data referenced by nested within the data 

This difference means that as I traverse through the composable controls I have to keep track of the key to reference nested data.

I am rethinking the call signature now though. I think it makes more sense to have the key as the last param.

Reply:
I am happy to report that I am wrong.  Details to come.